### PR TITLE
Separate trusted peers from config.json

### PR DESCRIPTION
**Pull request by MizukiSonoko**
Signed-off-by: Sonoko Mizuki <sonoko@mizuki.io>

### CODE REVIEW

File: iroha/src/config.rs [hunk 1]
```diff
@@ -91,6 +91,19 @@ impl Configuration {
...
+        path: P,
+    ) -> Result<(), String> {
+        let file = File::open(path).map_err(|e| format!("Failed to open a file: {}", e))?;
+        let reader = BufReader::new(file);
+        let trusted_peers: BTreeSet<PeerId> = serde_json::from_reader(reader)
+            .map_err(|e| format!("Failed to deserialize json from reader: {}", e))?;
+        self.sumeragi_configuration.trusted_peers = trusted_peers;
+        Ok(())
+    }
+
```
Comment (humb1t) [inline iroha/src/config.rs:106]: If we extracted trusted peers configuration from main configuration file, would be better to create an additional structure to load and store this new configuration.

Comment (MizukiSonoko) [reply iroha/src/config.rs:106]: @humb1t
Is it means separate `trusted peers` from `sumeragi configuration`? or Make the structure of trusted peers and is used when only load config as follows?
```
let trusted_peers: TrustedPeers = serde_json::from_reader(reader)
    .map_err(|e| format!("Failed to deserialize json from reader: {}", e))?;
self.sumeragi_configuration.trusted_peers = trusted_peers;
```

Comment (humb1t) [reply iroha/src/config.rs:106]: @MizukiSonoko I think second option is good enough.

Comment (eadventurous) [reply iroha/src/config.rs:106]: Or something like this, I think
```rust
let trusted_peers = TrustedPeers::from_path(&path)?;
self.sumeragi_configuration.trusted_peers = trusted_peers;
```

**Review by eadventurous — COMMENTED** [review]
I suggest to also add logs
```rust
log::info("Loading trusted peers from path: {}", path);
// ...
log::info("Trusted peers: {}", truested_peers);
```

File: iroha/src/config.rs (hunk 1 above)
Comment (humb1t) [reply iroha/src/config.rs:106]:
```
self.sumeragi_configuration.trusted_peers = TrustedPeers::from_path(&path)?;
```
=)

File: iroha/src/config.rs [hunk 2]
```diff
@@ -92,6 +92,15 @@ impl Configuration {
         Ok(())
     }
 
+    /// Load trusted peers variables from a json *pretty* formatted file.
+    pub fn load_trusted_peers_from_path<P: AsRef<Path> + Debug>(
+        &mut self,
+        path: P,
+    ) -> Result<(), String> {
+        self.sumeragi_configuration.trusted_peers = TrustedPeers::from_path(&path)?;
```
Comment (i1i1) [inline iroha/src/config.rs:103]: May be `Configuration::from_path` can take 2 arguments one of which can be path of config and other path of trusted peers?

File: iroha/src/sumeragi.rs [hunk 3]
```diff
@@ -123,209 +119,122 @@ impl Sumeragi {
...
-                    send_futures.push(message.clone().send_to(peer));
-                }
+                });
             }
             let results = futures::future::join_all(send_futures).await;
             results
-                .into_iter()
-                .filter_map(|result| result.err())
+                .iter()
+                .filter(|result| result.is_err())
```
Comment (i1i1) [inline iroha/src/sumeragi.rs:234]: Here `filter_map(|result| result.err())` would suit better, as it filters Ok-es and unwraps error

File: iroha/src/sumeragi.rs [hunk 4]
```diff
@@ -499,49 +385,14 @@ impl NetworkTopology {
...
 
 /// Represents a topology of peers, defining a `role` for each peer based on the previous block hash.
-#[derive(Debug, Clone, Encode, Decode)]
+#[derive(Debug, Clone)]
 pub struct InitializedNetworkTopology {
     /// Current order of peers. The roles of peers are defined based on this order.
     pub sorted_peers: Vec<PeerId>,
-    /// Maximum faulty peers in a network.
-    pub max_faults: u32,
+    max_faults: usize,
```
Comment (i1i1) [inline iroha/src/sumeragi.rs:392]: Should field here be public?

File: iroha/src/sumeragi.rs [hunk 5]
```diff
@@ -123,209 +119,122 @@ impl Sumeragi {
...
+                Ok(())
+            }
         } else {
-            self.forward_transactions_to_leader(&transactions).await?;
-            self.gossip_transactions(&transactions).await
-        }
-    }
-
-    /// Forwards transactions to the leader and waits for receipts.
-    pub async fn forward_transactions_to_leader(
```
Comment (i1i1) [inline iroha/src/sumeragi.rs:173]: May be returning `forward_transactions_to_leader` and `gossip_transactions` would increase readability

File: iroha/src/sumeragi.rs [hunk 6]
```diff
@@ -123,209 +119,122 @@ impl Sumeragi {
...
+                let mut send_futures = Vec::new();
+                for peer in &recipient_peers {
+                    if self.peer_id != *peer {
+                        send_futures.push(message.clone().send_to(peer));
+                    }
+                }
+                let results = futures::future::join_all(send_futures).await;
+                results
+                    .iter()
+                    .filter(|result| result.is_err())
```
Comment (i1i1) [inline iroha/src/sumeragi.rs:160]: Here `filter_map(|result| result.err())` would suit better, as it filters Ok-es and also unwraps errors

File: iroha/src/sumeragi.rs [hunk 7]
```diff
@@ -1437,15 +1199,40 @@ pub mod message {
...
+                                .expect("Failed to sign."),
+                        );
+                        let mut send_futures = Vec::new();
+                        for peer in &sumeragi.network_topology.sorted_peers {
+                            send_futures.push(message.clone().send_to(peer));
+                        }
+                        let results = futures::future::join_all(send_futures).await;
+                        results
+                            .iter()
+                            .filter(|result| result.is_err())
```
Comment (i1i1) [inline iroha/src/sumeragi.rs:1227]: `filter_map` would be better here

File: iroha/src/sumeragi.rs [hunk 8]
```diff
@@ -1964,12 +1746,12 @@ mod tests {
                 sumeragi.clone(),
                 (events_sender.clone(), events_receiver),
             );
-            let _ = task::spawn(async move {
+            task::spawn(async move {
```
Comment (i1i1) [inline iroha/src/sumeragi.rs:1749]: Clippy would warn about unused result here as task::spawn returns `JoinHandle`

**Review by humb1t — APPROVED** [review]

**Review by eadventurous — CHANGES_REQUESTED** [review]
I think this PR was improperly rebased/merged, `sumeragi.rs` definitely reverts some changes here, which it should not...

**Pull request closed without merging**